include/device.h

```diff
@@ -439,6 +443,70 @@ int device_any_busy_check(void);
  */
 int device_busy_check(struct device *chk_dev);
 
+/**
+ * @brief Init device idle PM
+ *
+ * @param dev Pointer to device structure of the specific device driver
+ * the caller is interested in.
+ */
+void device_ipm_init(struct device *dev);
+
+/**
+ * @brief Enable device idle PM
+ *
+ * @param dev Pointer to device structure of the specific device driver
+ * the caller is interested in.
+ */
+void device_ipm_enable(struct device *dev);
+
+/**
+ * @brief Disable device idle PM
+ *
+ * @param dev Pointer to device structure of the specific device driver
+ * the caller is interested in.
+ */
+void device_ipm_disable(struct device *dev);
+
+/**
+ * @brief Call device resume asynchronously based on usage count
+ *
+ * @param dev Pointer to device structure of the specific device driver
+ * the caller is interested in.
+ * @retval 0 If successful.
+ * @retval Errno Negative errno code if failure.
+ */
+int device_ipm_get(struct device *dev);
```

Comment:
General comments:
- The asynchronous interface needs a callback to notify when device is ready to be used or when it entered the idle mode. It also has to support transitions triggered in the middle of other transitions (one thread called put() and other get() almost simultaneously but device needs XXXms to do transition).
- The asynchronous calls needs to trace multiple users due to separate callbacks (there could be several subsystems requesting to either enable or disable power saving).

Reply:
we can add notification mechanism for async calls and pass the error no to let the driver know the status.
what do you mean by "trace multiple users"?

Reply:
Please consider device which needs some time to enter sleep mode or wake-up (I2C-connected gyroscope for example). The gyro is used by 2 mutually exclusive functionalities. What will happen if one functionality will be enabled just after disabling another?

Func1: Disabled
-> Gyro is now idle, we can go to suspend state.
-> Driver sends SUSPEND command (to be more specific schedules an I2C transaction).
-> Now driver waits for I2C transaction result, so other things can be scheduled.

Func2: Enabled
-> What is state of the Gyro? Should we send Wakeup command?

If Func1 and Func2 are using blocking calls, the lock on the device will solve this problem, however the Func1 will sleep with lock held . If one of this functionality would like to use asynchronous API, we have to trace the state of the device in the PM infrastructure.